src/Dimensions.cpp

```diff
@@ -221,8 +221,10 @@ SampledDimension& SampledDimension::operator=(const Dimension &other) {
 
     if (other.dimensionType() == DimensionType::Sample && impl() != tmp) {
         std::swap(impl(), tmp);
+    } else {
+        throw nix::IncompatibleDimensions("Cannot assign dimension of type " + nix::util::dimTypeToStr(other.dimensionType())
```

Comment:
Small thing: Now the exception in the else case will be thrown if it either not the right type, or if we have self-assignment.